[windows] redo windows tailer as polling rather than async.

derekwbrown opened the pull request:
### What does this PR do?

adds support for logs (windows)

derekwbrown committed:
[windows] redo windows tailer as polling rather than async.

tmichelet commented on pkg/logs/input/tailer/tailer.go:
```diff
@@ -45,7 +42,7 @@ type Tailer struct {
 	shouldStop   bool
 	stopTimer    *time.Timer
-	stopMutex    sync.Mutex
+	stopMutex    sync.Mutex;
```
nit: `;`

tmichelet commented on pkg/logs/input/tailer/tailer_nix.go:
```diff
@@ -0,0 +1,101 @@
+		return true, nil
+	}
+	return false, nil
+}
```
nit: no new line

tmichelet commented on pkg/logs/input/tailer/tailer_nix.go:
```diff
@@ -0,0 +1,101 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+// +build !windows
```
agent doesn't compile on unix:
```
# github.com/DataDog/datadog-agent/pkg/logs/input/tailer
pkg/logs/input/tailer/tailer_nix.go:11:2: imported and not used: "fmt"
pkg/logs/input/tailer/tailer_nix.go:15:2: imported and not used: "sync"
pkg/logs/input/tailer/tailer_nix.go:16:2: imported and not used: "sync/atomic"
pkg/logs/input/tailer/tailer_nix.go:17:2: imported and not used: "time"
pkg/logs/input/tailer/tailer_nix.go:21:2: imported and not used: "github.com/DataDog/datadog-agent/pkg/logs/config"
pkg/logs/input/tailer/tailer_nix.go:23:2: imported and not used: "github.com/DataDog/datadog-agent/pkg/logs/message"
pkg/logs/input/tailer/tailer_nix.go:93:5: undefined: inode
```

Happy to fix it myself if you don't have a working unix environment

tmichelet commented on pkg/logs/input/tailer/tailer_nix.go:
```diff
@@ -0,0 +1,101 @@
+	}
+}
+func (t *Tailer) checkForRotation() (bool, error) {
+	f, err := os.Open(t.path)
```
this method has been updated on the master branch, see https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/input/tailer/scanner.go#L151-L170

Could you port the updates here?

tmichelet reviewed (commented):
Thanks Derek, works as expected with little forking from unix!

Could you address the small points, remove the debug logs and comments?

Also, I think we can make `checkForRotation` very close on windows and on unix, so let's try to implement that too

derekwbrown committed:
Add reno notes

tmichelet reviewed (commented):
Getting close! Added a bunch of small comments on this part of the code.

Also, one feature we have is storing the offset of the last log line we tailed and starting again from there when the agent restarts. This currently assumes that a line ends with `\n`, which is not exactly true on windows. A quick fix is to add a EOL_SIZE constant and update these two lines:
- https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/decoder/line_handler.go#L81
- https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/decoder/line_buffer.go#L39

Could you have a look at that?

~Lastly, when testing this I noticed a bug in the logic to store the offset, when the log line ends with a whitespace we trim it and store the wrong offset. Let's not fix this right now but keep that in mind in case you end up in this edge case in your own testing.~ This bug is fixed by #1095

tmichelet committed:
[Logs] Windows tailing - add comments

tmichelet committed:
[Logs] Windows tailing - fix issue with offset storage

as we don't stop and start the tailer, the decoder offset is wrong.
A proper fix would be to use the same log rotation logic,
but given time constraints we'll go for the hack for now

tmichelet commented:
Update: I was wrong in my comment, there was indeed an issue but it was unrelated to the size of EOL, and is fixed by last commit

ajacquemot commented on pkg/logs/input/tailer/tailer_windows.go:
```diff
@@ -0,0 +1,107 @@
+
+	for {
+		n, err := f.Read(inBuf)
+		if n == 0 || err != nil {
```
I think you should return the error to catch `EOF`

tmichelet replied:
good point

ajacquemot commented on pkg/logs/input/tailer/tailer_windows.go:
```diff
@@ -0,0 +1,107 @@
+			t.SetDecodedOffset(0)
+		}
+	} else {
+		log.Debugf("Error stat()ing file %v", err)
```
What does it mean to get an error on file stats ? Don't we want to return this error ?

tmichelet replied:
Not sure, but it's not 100% required for the feature so we'll go ahead

tmichelet committed:
review fixes

derekwbrown committed:
Fix appveyor problem
